### ISSUE TITLE
Getting nillable parameters from EXP

### ISSUE THREAD

**Issue by femesq** [buildingSMART/IfcDoc#67]
I've got an validation error in the IFC line below:
```
#677826=IFCCLOSEDSHELL(());
```

Checking the `EXP`press spec, I could see that `CfsFaces` is a `SET` of [1..?] length:
![image](https://user-images.githubusercontent.com/1899668/89138840-42cc4680-d513-11ea-8c46-3aed0ffa37d6.png)

In this case, the IFC invalidation would be correct, but checking the XSD it seems to be optional (code below from [element's parent](https://standards.buildingsmart.org/IFC/RELEASE/IFC4/FINAL/HTML/schema/ifctopologyresource/lexical/ifcconnectedfaceset.htm)):
![image](https://user-images.githubusercontent.com/1899668/89138965-b0787280-d513-11ea-8d94-c8a5331b799c.png)

Verifying similar situations (like [this](https://standards.buildingsmart.org/IFC/RELEASE/IFC4/FINAL/HTML/schema/ifckernel/lexical/ifccomplexpropertytemplate.htm)), the `EXP`ress spec explicitly says the parameter is optional, matching XSD spec:
![image](https://user-images.githubusercontent.com/1899668/89139261-85dae980-d514-11ea-987d-08acff956447.png)

Am I missing something?

**Comment by aothms** [comment]
@jmirtsch told me that most of the XSD attributes are optional because of the fact that it allows instances to be added directly as child nodes with attributes, but also allows references to elements in other places of the tree, in which case there would not be any attributes but rather a node of the same entity type with a uri ref. This renders the XSD pretty much pointless for validation using out of the box tools.

So, in short. The EXPRESS schema is leading.

**Comment by jmirtsch** [comment]
This is my understanding of the situation.  
I'd love to see the xsd improved so XML validation could be utilized, but it requires XSD expertise beyond my current level (or time to research).

**Comment by TLiebich** [comment]
yes - when developing the (at that time) new XSD generation for IFC4, we stumpled over an issue, that could not been solved at that time. An element (entity in EXPRESS) can be used by containment and by reference (ID/REF) in XSD. But when using it as reference, it shall be empty - i.e. it should not have additional attributes. But this refers to EXPRESS attributes, like globalid which is required in EXPRESS but an optional attribute in XSD. An example.

If the GlobalId Attribute would map into a required attribute in XSD, it would have to be present in all cases, the element would be used in a reference. Like here, IfcBeam is used by reference:

```
<IfcRelAssociatesMaterial GlobalId="0e9JZKFZHE1xGhwKv0kdNS">
  <OwnerHistory xsi:nil="true" href="i1"/>
  <RelatedObjects>
    <IfcBeam xsi:nil="true" href="i67" GlobalId="0HgX8eMdT5IAlhYvOpq09J"/>
  </RelatedObjects>
  <RelatingMaterial>
    <IfcMaterial xsi:nil="true" href="i34"/>
  </RelatingMaterial>
</IfcRelAssociatesMaterial>

<IfcBeam id="i67" GlobalId="0HgX8eMdT5IAlhYvOpq09J" Name="BMR - 001">
  ...
</IfcBeam>
```
this causes an error, because when xsi:nill is true, there should be no content.
if there is a good solution for this issues, please forward.


The issue from @femesq is different however. Here it is (if I understand correctly) about the missing validation in XSD, that a valid closed shell should have at least one fase as part of the set of CfsFaces. The following instance is incorrect:

`#677826=IFCCLOSEDSHELL(());`

in my view, the xsd is correct here:

```
<xs:complexType name="IfcConnectedFaceSet">
  <xs:complexContent>
    <xs:extension base="ifc:IfcTopologicalRepresentationItem">
      <xs:sequence>
        <xs:element name="CfsFaces">  <!-- a single element of CfsFaces - so it is not optional her -->
          <xs:complexType>
            <xs:sequence>
              <xs:element ref="ifc:IfcFace" maxOccurs="unbounded"/>  <!-- minOccurs (not present) defaults to 1, max to N so it is correct as in SET[1:?]
            </xs:sequence>
            <xs:attribute ref="ifc:itemType" fixed="ifc:IfcFace"/>
            <xs:attribute ref="ifc:cType" fixed="set"/>
            <xs:attribute ref="ifc:arraySize" use="optional"/>  <!-- this is a special attribute required in ISO10303-28, and only to be used, if the array size is fixed, in this case it is unbounded, so it is not needed and does not influence any validation. -->
          </xs:complexType>
        </xs:element>
      </xs:sequence>
    </xs:extension>
  </xs:complexContent>
</xs:complexType>
```

not sure, why the compiler does not come up with an error here.

**Comment by aothms** [comment]
>  if there is a good solution for this issues, please forward.

I'd say one possible solution (and I'm not sure if this is possible in the framework of p28) is to create three (!) XSD types for every entity. Let's say `IfcBeamDef` (for an inline definition) and `IfcBeamRef` (for reference) and with `IfcBeam` a union of `IfcBeamDef` and `IfcBeamRef`. But it will obviously create a whole lotta cruft, although the added types are basically all identical.

**Comment by femesq** [comment]
In that case, it's possible a problem with older version of the authoring tool...
```
FILE_DESCRIPTION(('ViewDefinition [PresentationView]'),'2;1');
FILE_NAME('SD-M-1637-L55-IFC 200318.IFC','2012-01-01T09:00:00',('User'),('Autodesk',''),'v3','XP SP2','');
```

Thank you everyone for putting an effort on pointing me the directions!

**Comment by SergejMuhic** [comment]
Hi Felipe,

older IFC files from Revit on occasion had their issues with geometry having empty lists. Quite awhile ago I have reported the issue and it was corrected.

The file you are working on appears to be from before the fix I mentioned above. Unless you have the original Revit file (and Revit since it seems too old to be supported) and try to generate the IFC file from it that geometry apparently just did not get exported and you will not be able to retrieve it.

If you would like, you can send me the file and I can take a look. Maybe it is just a reference that is missing. This was also an issue with older Revit exports where you could just manually find the entities and reference them in the owning class.

**Comment by femesq** [comment]
@SergejMuhic, thanks for your availability.
Unfortunately, I don't dave the Revit file and can't send the IFC (it's from a customer), but I can confirm it's old-version (2012) Revit exported.
[]'s
